python3Packages.plyer: init at 2.0.0

rski opened the pull request:
This is a dependency of the newer versions of mirage-im.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

rski committed:
python3Packages.plyer: init at 2.0.0

This is a dependency of the newer versions of mirage-im.

rski commented:
took a stab at packaging this to upgrade mirage, found out afterworks that there is an attempt open, but it seems stalled. adding sandro and vojta001 because of the comments in https://github.com/NixOS/nixpkgs/pull/115189/commits/90df4f769af8a1ffe2c91e55a73046d5c443d0d9

risicle commented:
Fails on macos 10.15:

```
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.8-plyer-2.0.0.drv-0/source/plyer/tests/common.py", line 53, in platform_import
    mod = __import__(module, fromlist='.')
  File "/private/tmp/nix-build-python3.8-plyer-2.0.0.drv-0/source/plyer/platforms/macosx/audio.py", line 3, in <module>
    from pyobjus import autoclass
ModuleNotFoundError: No module named 'pyobjus'
E.upower not found.upower not found....ioreg not found....system_profiler not found............{}
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.8-plyer-2.0.0.drv-0/source/plyer/tests/common.py", line 53, in platform_import
    mod = __import__(module, fromlist='.')
  File "/private/tmp/nix-build-python3.8-plyer-2.0.0.drv-0/source/plyer/platforms/macosx/screenshot.py", line 5, in <module>
    from plyer.platforms.macosx.storagepath import OSXStoragePath
  File "/private/tmp/nix-build-python3.8-plyer-2.0.0.drv-0/source/plyer/platforms/macosx/storagepath.py", line 7, in <module>
    from pyobjus import autoclass
ModuleNotFoundError: No module named 'pyobjus'
E.{}
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.8-plyer-2.0.0.drv-0/source/plyer/tests/common.py", line 53, in platform_import
    mod = __import__(module, fromlist='.')
  File "/private/tmp/nix-build-python3.8-plyer-2.0.0.drv-0/source/plyer/platforms/macosx/storagepath.py", line 7, in <module>
    from pyobjus import autoclass
ModuleNotFoundError: No module named 'pyobjus'
E.s..........
======================================================================
ERROR: test_audio_macosx (plyer.tests.test_audio.TestAudio)
Test macOS audio start, stop and play
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.8-plyer-2.0.0.drv-0/source/plyer/tests/test_audio.py", line 40, in test_audio_macosx
    audio = platform_import(
  File "/private/tmp/nix-build-python3.8-plyer-2.0.0.drv-0/source/plyer/tests/common.py", line 61, in platform_import
    return mod
UnboundLocalError: local variable 'mod' referenced before assignment

======================================================================
ERROR: test_screenshot_screencapture (plyer.tests.test_screenshot.TestScreenshot)
Test mocked MacOS screencapture for plyer.screenshot.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.8-plyer-2.0.0.drv-0/source/plyer/tests/test_screenshot.py", line 90, in test_screenshot_screencapture
    scr = platform_import(
  File "/private/tmp/nix-build-python3.8-plyer-2.0.0.drv-0/source/plyer/tests/common.py", line 60, in platform_import
    mod.whereis_exe = whereis_exe
UnboundLocalError: local variable 'mod' referenced before assignment

======================================================================
ERROR: test_storagepath_macosx (plyer.tests.test_storagepath.TestStoragePath)
Test macOS for plyer.storagepath.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/tmp/nix-build-python3.8-plyer-2.0.0.drv-0/source/plyer/tests/test_storagepath.py", line 25, in test_storagepath_macosx
    storagepath = platform_import(
  File "/private/tmp/nix-build-python3.8-plyer-2.0.0.drv-0/source/plyer/tests/common.py", line 61, in platform_import
    return mod
UnboundLocalError: local variable 'mod' referenced before assignment

----------------------------------------------------------------------
Ran 34 tests in 0.091s

FAILED (errors=3, skipped=1)
```

I'm thinking its dependencies vary significantly by platform.

rski commented:
Whoops, yes. Forgot to mark it as broken, i don't have access to macos to test it. Mirage is currently broken on macos too.

legendofmiracles commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,75 @@
+  src = fetchFromGitHub {
+    owner = "kivy";
+    repo = "plyer";
+    rev = "${version}";
```
```suggestion
    rev = version;
```

samuela commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,77 @@
+{ lib, buildPythonPackage, fetchFromGitHub, mock, python3Packages, python, stdenv }:
+
+buildPythonPackage rec {
+
```
```suggestion
```

samuela commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,77 @@
+  '';
+
+  postPatch = ''
+    # Fixup this file to
+    # - not assume ~/.config/user-dirs.dirs exists
+    # - not assume it actually has stuff in it if it does exist
+    # - return the path under HOME if user-dirs has no useful info.
+    # - use the right name for DOCUMENTS according to the spec
```
Are there upstream issues for these bugs? If so it would be nice to include them. If not, we should probably report them.

rski replied:
there is a pull request open for these

samuela replied:
We should link to those in the comment so that people reading in the future can have a paper trail

samuela replied:
Also if possible we should actually just patch with that pr using fetchpatch instead of manually doing it with a bunch of string replaces

rski replied:
I submitted a PR and changed it to fetch the patches from that PR

SuperSandro2000 commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,77 @@
+{ lib, buildPythonPackage, fetchFromGitHub, mock, python3Packages, python, stdenv }:
+
+buildPythonPackage rec {
+
+  pname = "plyer";
```
```suggestion
buildPythonPackage rec {
  pname = "plyer";
```

SuperSandro2000 commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,77 @@
+{ lib, buildPythonPackage, fetchFromGitHub, mock, python3Packages, python, stdenv }:
```
```suggestion
{ lib, buildPythonPackage, fetchFromGitHub, mock, python, stdenv }:
```
this is not allowed in python packages. Please list the packages individually.

SuperSandro2000 commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,77 @@
+    sha256 = "15z1wpq6s69s76r6akzgg340bpc21l2r1j8270gp7i1rpnffcjwm";
+  };
+
+  propagatedBuildInputs = [ python3Packages.keyring ];
```
```suggestion
  propagatedBuildInputs = [ keyring ];
```

SuperSandro2000 commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,77 @@
+  checkInputs = [ mock ];
+  preCheck = ''
+    HOME=$(mktemp -d)
+    mkdir -p $HOME/.config/
+    touch $HOME/.config/user-dirs.dirs
+    mkdir -p $HOME/Pictures/
```
```suggestion
    mkdir -p $HOME/.config $HOME/Pictures
    touch $HOME/.config/user-dirs.dirs
```

SuperSandro2000 commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,77 @@
+  '';
+  checkPhase = ''
+    runHook preCheck
+    ${python.interpreter} -m unittest discover \
+      --start-directory plyer/tests \
+      --top-level-directory .
```
Can this be done with pytest?

samuela replied:
and pytestCheckHook if so?

SuperSandro2000 commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,77 @@
+    runHook postCheck
+  '';
+
+  postPatch = ''
```
This belongs after src.

SuperSandro2000 commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,77 @@
+    # - return the path under HOME if user-dirs has no useful info.
+    # - use the right name for DOCUMENTS according to the spec
+    substituteInPlace plyer/platforms/linux/storagepath.py \
+      --replace \
+        "except KeyError:" "except (KeyError, IndexError, FileNotFoundError):" \
+      --replace \
+        "return PATHS[name]" "return self._get_home_dir() + PATHS[name]" \
+      --replace \
+        $'directory = self._get_from_user_dirs("DOCUMENT")' \
+        $'directory = self._get_from_user_dirs("DOCUMENTS")'
```
```suggestion
      --replace "except KeyError:" "except (KeyError, IndexError, FileNotFoundError):" \
      --replace "return PATHS[name]" "return self._get_home_dir() + PATHS[name]" \
      --replace $'directory = self._get_from_user_dirs("DOCUMENT")' \
        $'directory = self._get_from_user_dirs("DOCUMENTS")'
```

SuperSandro2000 commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,77 @@
+
+    rm -r examples
+    # Test assumes dbus is not installed, it fails and is not very robust.
+    substituteInPlace plyer/tests/test_notification.py --replace \
+      "    def test_notification_notifysend(self):" \
+      $'    @unittest.skip("")\n    def test_notification_notifysend(self):'
```
```suggestion
    substituteInPlace plyer/tests/test_notification.py \
      --replace "def test_notification_notifysend(self):" \
        $'@unittest.skip("")\n    def test_notification_notifysend(self):'
```

SuperSandro2000 commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,77 @@
+      $'    @unittest.skip("")\n    def test_notification_notifysend(self):'
+    # this fails to produce a uniqueid.id, but I would not be able to explain why.
+    # It does work outside the nix build.
+    substituteInPlace plyer/tests/test_uniqueid.py --replace \
+      "    def test_uniqueid(self):" \
+      $'    @unittest.skip("")\n    def test_uniqueid(self):'
```
```suggestion
    substituteInPlace plyer/tests/test_uniqueid.py \
      --replace "def test_uniqueid(self):" \
        $'@unittest.skip("")\n    def test_uniqueid(self):'
```

SuperSandro2000 commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,77 @@
+  pythonImportsCheck = [ "plyer" ];
+
+  meta = with lib; {
+    description =
+      "Plyer is a platform-independent api to use features commonly found on various platforms";
```
```suggestion
    description = "Plyer is a platform-independent api to use features commonly found on various platforms";
```

SuperSandro2000 commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,77 @@
+    homepage = "https://github.com/kivy/plyer";
+    license = licenses.mit;
+    maintainers = with maintainers; [ rski ];
+    broken = stdenv.isDarwin;
```
This should probably be converted into a disabled if it never will work on darwin.

rski replied:
the plyer package is supposed to provide cross platform APIs that include darwin. I don't have access to macos hardware to fix it though. If anyone wants to pick it up they can though.

SuperSandro2000 replied:
```suggestion
```
If there is no concrete build failure and you just did not test it on darwin we don't set this flag.

rski replied:
it definitely is broken on Darwin, see https://github.com/NixOS/nixpkgs/pull/136902#issuecomment-913862312

rski commented:
ok, the comments left so far should be addressed now

samuela commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,60 @@
+  # notifysend assumes dbus is not installed, it fails and is not very robust.
+  # test_uniqueid fails during nix-build, but I am not able to explain why. 
+  # The test and the API under test do work outside the nix build.
+  pytestFlagsArray = [ "plyer/tests" ''-k "not test_notification_notifysend and not test_uniqueid"'' ];
```
You should be able to use disabledTests instead of -k in pytestFlagsArray

rski replied:
oh nice, done

samuela commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,64 @@
+
+  src = fetchFromGitHub {
+    owner = "kivy";
+    repo = "plyer";
```
```suggestion
    repo = pname;
```

rski replied:
whoops, did I accidentally resolve this earlier? Fixed now

SuperSandro2000 committed:
Update pkgs/development/python-modules/plyer/default.nix

SuperSandro2000 commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,63 @@
+    # fix handling of the ~/.config/user-dirs.dir file
+    (fetchpatch {
+      url = "https://github.com/rski/plyer/commit/f803697a1fe4fb5e9c729ee6ef1997b8d64f3ccd.patch";
+      sha256 = "1ciksqfn63bpn9p3s1min24ipjqp32pgcsffabkxj2b5d1nnbplz";
```
```suggestion
      sha256 = "sha256-akuh//P5puz2PwcBRXZQ4KoGk+fxi4jn2H3pTIT5M78=";
```

SuperSandro2000 commented on pkgs/development/python-modules/plyer/default.nix:
```diff
@@ -0,0 +1,63 @@
+    # fix naming of the DOCUMENTS dir
+    (fetchpatch {
+      url = "https://github.com/rski/plyer/commit/99dabb2d62248fc3ea5705c2720abf71c9fc378b.patch";
+      sha256 = "1ygahxpb8l6bqwpfba0cm164mwyc2snhpdighzyw540s1l0n31g8";
```
```suggestion
      sha256 = "sha256-bbnw0TxH4FGTso5dopzquDCjrjZAy+6CJauqi/nfstA=";
```

SuperSandro2000 reviewed (changes requested):
How did you generate those hashes?

rski commented:
> How did you generate those hashes?

nix-prefetch-url, is something wrong with them?

SuperSandro2000 commented:
> > How did you generate those hashes?
> 
> nix-prefetch-url, is something wrong with them?

Yeah, that is wrong. fetchpatch does some normalisation so they need to be updated.

rski committed:
python3Packages.plyer: use correct hashes for patches

nix-prefetch-url works differently to fetchpatch, use the correct hashes
here.

Co-authored-by: Sandro <sandro.jaeckel@gmail.com>

SuperSandro2000 commented:
It is not about that the hashes are SRI or not. That does not matter at all. The hashes required for fetchpatch are calculated differently that nix-prefetch-url.

rski commented:
weee, thank you all for your patience and help
